File: sdk/typescript/src/client/sensitive.ts
/**
 * SensitiveDataValidator (§1.8) — credentials must travel via env, never in a
 * register body or task parameters.
 *
 * Scan an object's keys against SENSITIVE_KEYWORDS; a sensitive key with a
 * non-empty value is a leak. The register path throws (fail-fast at start);
 * the task-params path returns a SECURITY_REJECTED signal so the runtime can
 * report the task failed instead of crashing the worker.
 */

import { SENSITIVE_KEYWORDS } from "../constants.ts";
import { ErrorCode } from "../protocol.ts";

/** Error thrown on the register path when a credential leaks into the body. */
export class SensitiveDataError extends Error {
  readonly leakedKeys: string[];
  constructor(leakedKeys: string[]) {
    super(
      `sensitive data detected in body; credentials must use env not payload: ${leakedKeys.join(
        ", ",
      )}`,
    );
    this.name = "SensitiveDataError";
    this.leakedKeys = leakedKeys;
  }
}

/** Outcome of a task-params scan. */
export interface SensitiveScanResult {
  ok: boolean;
  /** Present only when !ok. */
  errorCode?: ErrorCode;
  leakedKeys: string[];
}

function normalizeKey(key: string): string {
  // case-insensitive, ignore separators so "API-Key" matches "apikey"
  return key.toLowerCase().replace(/[_\-\s]/g, "");
}

export class SensitiveDataValidator {
  #keywords: Set<string>;

  /** @param extraKeywords tenant deny-list extension (§1.8 hook). */
  constructor(extraKeywords: readonly string[] = []) {
    this.#keywords = new Set(
      [...SENSITIVE_KEYWORDS, ...extraKeywords].map(normalizeKey),
    );
  }

  /** Extend the deny-list at runtime (returns this for chaining). */
  addKeyword(keyword: string): this {
    this.#keywords.add(normalizeKey(keyword));
    return this;
  }

  /**
   * Collect keys whose normalized form matches the deny-list AND whose value is
   * a non-empty string. Empty / null / undefined values are allowed (the field
   * may legitimately exist as a placeholder).
   */
  #scan(obj: unknown): string[] {
    const leaked: string[] = [];
    if (obj == null || typeof obj !== "object") return leaked;
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      const norm = normalizeKey(key);
      const matched = [...this.#keywords].some(
        (kw) => norm === kw || norm.includes(kw),
      );
      if (!matched) continue;
      if (typeof value === "string" && value.length > 0) {
        leaked.push(key);
      } else if (typeof value === "number" || typeof value === "boolean") {
        // a non-string truthy credential is still a leak
        leaked.push(key);
      }
    }
    return leaked;
  }

  /** Register path: throw on any leak (fail-fast at worker start). */
  assertRegisterBody(body: unknown): void {
    const leaked = this.#scan(body);
    if (leaked.length > 0) {
      throw new SensitiveDataError(leaked);
    }
  }

  /** Task-params path: return SECURITY_REJECTED instead of throwing. */
  scanTaskParams(params: unknown): SensitiveScanResult {
    const leaked = this.#scan(params);
    if (leaked.length > 0) {
      return {
        ok: false,
        errorCode: ErrorCode.SECURITY_REJECTED,
        leakedKeys: leaked,
      };
    }
    return { ok: true, leakedKeys: [] };
  }
}

File: sdk/go/client/sensitive.go
package client

import (
	"fmt"
	"strings"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/protocol"
)

// SensitiveValidator implements the SensitiveDataValidator equivalent
// (byo-sdk-guide §1.8): credentials must travel via env, never in a register
// body or dispatch parameters. It scans map keys against
// protocol.SensitiveKeywords (plus a tenant-extensible deny-list); a sensitive
// key carrying a NON-EMPTY value is a violation.
//
// Two enforcement points differ by phase:
//   - register body  -> ValidateRegister returns an error (fail-fast at startup).
//   - dispatch params -> ValidateParameters returns SECURITY_REJECTED so the
//     task is reported failed rather than crashing the worker.
type SensitiveValidator struct {
	// keywords is the effective deny-list (shared protocol set + extensions),
	// stored lower-cased and stripped of separators for matching.
	keywords []string
}

// NewSensitiveValidator builds a validator seeded with protocol.SensitiveKeywords
// and any tenant deny-list extensions.
func NewSensitiveValidator(extraDenyList ...string) *SensitiveValidator {
	v := &SensitiveValidator{}
	for _, k := range protocol.SensitiveKeywords {
		v.keywords = append(v.keywords, normalizeKey(k))
	}
	v.AddDenyList(extraDenyList...)
	return v
}

// AddDenyList registers additional sensitive keywords (the §1.8 extension hook).
func (v *SensitiveValidator) AddDenyList(keywords ...string) {
	for _, k := range keywords {
		n := normalizeKey(k)
		if n != "" && !containsString(v.keywords, n) {
			v.keywords = append(v.keywords, n)
		}
	}
}

// normalizeKey lower-cases and strips '_' / '-' / '.' so "API_KEY", "api-key"
// and "apikey" all collapse to the same token.
func normalizeKey(k string) string {
	k = strings.ToLower(k)
	k = strings.NewReplacer("_", "", "-", "", ".", "").Replace(k)
	return k
}

// isSensitiveKey reports whether key matches any deny-list keyword (substring on
// the normalized form, matching the Java validator's contains semantics).
func (v *SensitiveValidator) isSensitiveKey(key string) bool {
	n := normalizeKey(key)
	for _, kw := range v.keywords {
		if strings.Contains(n, kw) {
			return true
		}
	}
	return false
}

// scan returns the first offending key whose value is sensitive + non-empty.
// Recurses into nested maps so parameters.credentials.password is caught.
func (v *SensitiveValidator) scan(m map[string]any) (string, bool) {
	for k, val := range m {
		if nested, ok := val.(map[string]any); ok {
			if bad, found := v.scan(nested); found {
				return bad, true
			}
			continue
		}
		if v.isSensitiveKey(k) && !isEmptyValue(val) {
			return k, true
		}
	}
	return "", false
}

// isEmptyValue reports whether a value should be treated as "blank" (allowed —
// credential goes via env, so an empty placeholder key is fine).
func isEmptyValue(val any) bool {
	switch x := val.(type) {
	case nil:
		return true
	case string:
		return strings.TrimSpace(x) == ""
	default:
		return false
	}
}

// ValidateRegister fails fast on any sensitive non-empty field in the register
// fingerprint attributes (§1.8 register path).
func (v *SensitiveValidator) ValidateRegister(req RegisterRequest) error {
	if bad, found := v.scan(req.Attributes); found {
		return fmt.Errorf("sensitive credential %q must travel via env, not the register body", bad)
	}
	return nil
}

// ValidateParameters checks dispatch effective parameters. A violation yields a
// SECURITY_REJECTED TaskResult (the second return is true when rejected) so the
// caller reports the task failed rather than executing it.
func (v *SensitiveValidator) ValidateParameters(params map[string]any) (TaskResult, bool) {
	if bad, found := v.scan(params); found {
		return Fail(protocol.ErrorCodeSecurityRejected,
			fmt.Sprintf("sensitive credential %q present in task parameters", bad)), true
	}
	return TaskResult{}, false
}

func containsString(s []string, t string) bool {
	for _, x := range s {
		if x == t {
			return true
		}
	}
	return false
}
